src/util/support/utf8_conv.c

```diff
@@ -430,7 +433,7 @@ krb5int_ucs2lecs_to_utf8s(const unsigned char *ucs2les,
     if (ucs2leslen > SSIZE_MAX)
         return ERANGE;
 
-    len = k5_ucs2s_to_utf8s(NULL, (krb5_ucs2 *)ucs2les, 0,
+    len = k5_ucs2s_to_utf8s(NULL, ucs2_of_utf8(ucs2les), 0,
```

Comment:
My bad; it looks like krb5int_ucs2lecs_to_utf8s is the only entry point into k5_ucs2s_to_utf8s whihc is used.

This conversion is unsafe because we don't know the alignment of the input pointer.  The byte swapping insid ek5_ucs2s_to_utf8s is also broken; if we ever asked for non-little-endian input on a little-endian platform, we would get the wrong values.  The upstream OpenLDAP code (in utf-8-conv.c) operates on a wchar_t array and does not do byte swapping, so this is all stuff introduced by lukeh when the code was adapted for krb5.

For what we want, I think the right answer is to:

* Remove the unused entry points, so that we are only trying to work on byte arrays in little-endian.

* Change k5_ucs2s_to_utf8s() to operate on a byte array and use load_16_le() to pull UCS-2 values out.  Probably also to use k5buf instead of operating in two passes and worrying about the output buffer size.

That's a fair amount of code surgery; I can take care of it if need be.